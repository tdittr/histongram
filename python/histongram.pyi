from typing import Iterator


class Histogram:
    def __init__(self): ...

    def add(self, key: str): ...

    def add_many(self, iter: Iterator[str]): ...

    def __getitem__(self, key: str) -> int: ...
