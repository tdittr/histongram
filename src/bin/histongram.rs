use std::fs::read_to_string;
use std::path::PathBuf;

use clap::Parser;
use color_eyre::Result;
use compact_str::CompactString;

use histongram::{Histogram, Ngrams};

#[derive(Parser, Debug, Clone)]
struct Args {
    file: PathBuf,

    #[clap(long)]
    print: bool,
}

fn main() -> Result<()> {
    color_eyre::install()?;
    let args = Args::parse();

    let data = read_to_string(args.file)?;

    let ngrams = Ngrams::new(1..=5).count(data.split_whitespace());

    if args.print {
        println!("{ngrams:?}");
    }

    Ok(())
}
