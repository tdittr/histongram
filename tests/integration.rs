use std::cmp::{min, Reverse};

use insta::assert_ron_snapshot;

use histongram::Histogram;

const APACHE: &str = include_str!("../LICENSE-APACHE");
const MIT: &str = include_str!("../LICENSE-MIT");

#[test]
fn license_words() {
    let mut a = Histogram::from_iter(APACHE.split_whitespace());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.split_whitespace());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[test]
fn license_chars() {
    let mut a = Histogram::from_iter(APACHE.chars());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.chars());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

fn sort_also_by_key<K: Ord + Copy>(counts: &mut [(K, usize)]) {
    // This table should be sorted by count. But items with the same number of occurrences are
    // in arbitrary order...
    let mut high = usize::MAX;
    for &(_, cnt) in counts.iter() {
        assert!(cnt <= high);
        high = min(high, cnt);
    }

    // Now sort also by key so the snapshot stays consistent
    counts.sort_unstable_by_key(|(key, cnt)| (Reverse(*cnt), *key));
}
