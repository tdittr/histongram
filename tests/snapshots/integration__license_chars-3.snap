---
source: tests/integration.rs
expression: combined_counts
---
[
  (' ', 2662),
  ('e', 879),
  ('t', 789),
  ('o', 778),
  ('i', 772),
  ('n', 645),
  ('r', 612),
  ('a', 543),
  ('s', 483),
  ('h', 327),
  ('c', 324),
  ('d', 299),
  ('l', 275),
  ('u', 256),
  ('
', 216),
  ('f', 193),
  ('m', 172),
  ('p', 167),
  (',', 163),
  ('y', 158),
  ('b', 147),
  ('g', 130),
  ('w', 104),
  ('I', 75),
  ('v', 73),
  ('T', 72),
  ('O', 69),
  ('N', 67),
  ('L', 65),
  ('k', 65),
  ('.', 63),
  ('E', 61),
  ('S', 61),
  ('W', 61),
  ('A', 54),
  ('R', 53),
  ('C', 47),
  ('"', 44),
  ('D', 44),
  ('Y', 42),
  ('(', 24),
  (')', 24),
  ('H', 23),
  ('F', 22),
  ('x', 21),
  ('U', 18),
  ('P', 15),
  ('-', 11),
  ('M', 11),
  ('/', 10),
  ('B', 9),
  ('G', 9),
  ('j', 9),
  ('0', 6),
  (':', 6),
  (';', 6),
  ('q', 6),
  ('2', 5),
  ('V', 4),
  ('K', 3),
  ('[', 3),
  (']', 3),
  ('1', 2),
  ('4', 2),
  ('5', 2),
  ('9', 2),
  ('X', 2),
  ('z', 2),
  ('!', 1),
  ('%', 1),
  ('\'', 1),
  ('3', 1),
  ('6', 1),
  ('7', 1),
  ('8', 1),
  ('J', 1),
]
